var/spack/repos/builtin/packages/mochi-margo/package.py

```diff
@@ -64,6 +64,10 @@ class MochiMargo(AutotoolsPackage):
     depends_on("mercury@1.0.0:", type=("build", "link", "run"), when="@:0.5.1")
     depends_on("mercury@2.0.0:", type=("build", "link", "run"), when="@0.5.2:")
 
+    # Fix pthread detection
+    # https://github.com/mochi-hpc/mochi-margo/pull/177
+    patch("mochi-margo-pthreads.patch", when="@0.9:0.9.7")
```

Comment:
One oddity: I tested this PR with the versions specified here (0.9 through 0.9.7) and noticed that it did not apply the patch for 0.9.  In my case I ran `spack install mochi-margo@0.9^mercury@2.1.0~boostsys` and observed `==> No patches needed for mochi-margo` in the output.

I thought the syntax on that should be inclusive of 0.9 in the version specification, though?

Otherwise the PR looks good; it is a clean backport of a feature test from newer versions of mochi-margo.

Reply:
Did it install `mochi-margo@0.9.8:`?

Reply:
There are three versions that spack will likely prefer with only `0.9` specified, `@0.9.8,0.9.9,0.9.10`

Reply:
Ah, good theory, but that's apparently not what's happening in my case:

```
> spack install mochi-margo@0.9^mercury@2.1.0~boostsys
...
==> Installing mochi-margo-0.9-plgextxtrk5dqqxz2iartkufzs4oalge
==> No binary for mochi-margo-0.9-plgextxtrk5dqqxz2iartkufzs4oalge found: installing from source
==> Using cached archive: /home/carns/working/src/spack/var/spack/cache/_source-cache/archive/a2/a24376f66450cc8fd7a43043e189f8efce5a931585e53c1e2e41894a3e99b517.tar.gz
==> No patches needed for mochi-margo
==> mochi-margo: Executing phase: 'autoreconf'
==> mochi-margo: Executing phase: 'configure'
==> mochi-margo: Executing phase: 'build'
==> mochi-margo: Executing phase: 'install'
==> mochi-margo: Successfully installed mochi-margo-0.9-plgextxtrk5dqqxz2iartkufzs4oalge
  Stage: 0.01s.  Autoreconf: 4.05s.  Configure: 3.95s.  Build: 2.27s.  Install: 0.29s.  Total: 10.67s
[+] /home/carns/working/src/spack/opt/spack/linux-ubuntu22.10-skylake/gcc-12.2.0/mochi-margo-0.9-plgextxtrk5dqqxz2iartkufzs4oalge
```

vs. what happens when I install 0.9.1:

```
> spack install mochi-margo@0.9.1^mercury@2.1.0~boostsys
...
==> Installing mochi-margo-0.9.1-l2jnqavyzoj57vsaszh4db37vaqdmpdm
==> No binary for mochi-margo-0.9.1-l2jnqavyzoj57vsaszh4db37vaqdmpdm found: installing from source
==> Using cached archive: /home/carns/working/src/spack/var/spack/cache/_source-cache/archive/3f/3fe933f2d758ef23d582bc776e4f8cfae9bf9d0849b8b1f9d73ee024e218f2bc.tar.gz
==> Applied patch /home/carns/working/src/spack/var/spack/repos/builtin/packages/mochi-margo/mochi-margo-pthreads.patch
==> mochi-margo: Executing phase: 'autoreconf'
==> mochi-margo: Executing phase: 'configure'
==> mochi-margo: Executing phase: 'build'
==> mochi-margo: Executing phase: 'install'
==> mochi-margo: Successfully installed mochi-margo-0.9.1-l2jnqavyzoj57vsaszh4db37vaqdmpdm
  Stage: 0.02s.  Autoreconf: 5.40s.  Configure: 3.87s.  Build: 4.65s.  Install: 0.41s.  Total: 14.47s
[+] /home/carns/working/src/spack/opt/spack/linux-ubuntu22.10-skylake/gcc-12.2.0/mochi-margo-0.9.1-l2jnqavyzoj57vsaszh4db37vaqdmpdm
```

It kind of doesn't matter (I doubt anyone is clamoring to install 0.9 specifically), but it is a little weird. Probably more of a general spack quirk than anything to do with this PR?

Reply:
yeah, that is actually  a bug in this package recipe, the versioning scheme is broken. I will fix those as well.

All it needs is to add a `.0` at the end of the initial major versions. Spack doesn't handle inconsistent versioning strings at all, it is not easy/possible to deal with that kind of thing reliably.

Reply:
It isn't consistent with the project tags, but now Spack will know what to do with them.

Reply:
1) We can start using x.y.z from now on in future versions (and in all Mochi libraries, for consistency);
2) For the present PR, we can simply write `patch(..., "@0.9,0.9.1:0.9.7")`, I believe that should work.

Changing the version numbers now means also changing them in our mochi-spack-packages repository.

Reply:
@kwryankrattiger as an aside, if there is a reason why unifyfs is sticking with an old version of margo, we would very much like to hear it, in case there is anything we can do about it.

Reply:
Just noticed this conversation while updating the unifyfs package.

> they have an explicit dependency on `mochi-margo@0.9.6`, but develop can us the rest of the 0.9 versions up to `@0.9.9`. Again, not sure why they are excluding `@0.9.10` here.

This restriction was put in when we discovered unifyfs failed to build/run with `@0.9.10` or later and we wanted a well tested version in place. Probably shouldn't have restricted it to just one version, but now the range for those versions has been increased to `@0.9.9`. The latest unifyfs release is now set to allow anything from `@0.10` on, although it's mostly been tested with `@0.13.1`.

> if there is a reason why unifyfs is sticking with an old version of margo, we would very much like to hear it, in case there is anything we can do about it.

No reason other than it simply took us a while to finally update unifyfs to work with the newer versions.